### PR TITLE
Strip email trackers mpp 1662

### PR DESCRIPTION
**Pull request by say-yawn**
This PR fixes #<issue ID>.

How to test:

- [ ] l10n changes have been submitted to the l10n repository, if any.
- [ ] I've added a unit test to test for potential regressions of this bug.
- [ ] I've added or updated relevant docs in the docs/ directory.
- [ ] All UI revisions follow the [coding standards](https://github.com/mozilla/fx-private-relay/blob/main/docs/coding-standards.md), and use Protocol tokens where applicable (see `/static/scss/libs/protocol/css/includes/tokens/dist/index.scss`).
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

# New feature description
## To create user group for internal users from Mozilla Corporation, Foundation, and Pocket, run Django commands
```
./manage.py get_or_create_user_group mozilla_corporation
./manage.py get_or_create_user_group mozilla_foundation
./manage.py get_or_create_user_group pocket
```
Check this returns:
```
SUCCESS: created mozilla_corporation with Group ID: 1.
SUCCESS: created mozilla_foundation with Group ID: 2.
SUCCESS: created pocket with Group ID: 3.
```

## To set the user with the appropriate group based on their email domain:
```
./manage.py update_user_group mozilla.com
./manage.py update_user_group mozillafoundation.org
./manage.py update_user_group getpocket.com
```
And check that for every command it returns
```
Updated {num} users group.
```

## To create flag for Foxfooding
```
./manage.py waffle_flag foxfood --group=mozilla_corporation --create
```
Check that it returns:
```
Creating flag: foxfood
Setting name: foxfood
Setting everyone: None
Setting percent: None
Setting superusers: False
Setting staff: False
Setting authenticated: False
Setting group(s): ['mozilla_corporation']
Setting user(s): set()
Setting rollout: False
```

## To add specific user on Foxfood flag
In psql run:
```
SELECT id FROM auth_user WHERE email='email@address.com';
```
Get the id of the user and in the project run:
```
./manage.py waffle_flag foxfood --append --user=email@address.com
```
Check this returns:
```
Setting name: foxfood
Setting everyone: None
Setting percent: None
Setting superusers: False
Setting staff: False
Setting authenticated: False
Setting group(s): ['mozilla_corporation']
Setting user(s): {<User: new-user>}
```
For some reason even when using the flag `--append` doing:
```
./manage.py waffle_flag foxfood --append --user=email@address.com
```
replaces the previously added group on the flag. So to preserve we need to do something like:
```
./manage.py waffle_flag foxfood --append --user=email@address.com --group=mozilla_corporation
```

# Screenshot (if applicable)

Not applicable.

# How to test
## Before creating flag, emails are relayed successfully
1. Before running the commands send email to relay address
2. Check that email was sent without a problem

## After running the flag command and adding the mozilla_corporation group, email relayed for mozilla.com user strips tracker and adds breakage survey link
1. To create a user group for Mozilla Corporation users and add users with mozilla.com email address, run:
```
./manage.py get_or_create_user_group mozilla_corporation
./manage.py update_user_group mozilla.com
```
2. To create a flag that is active for Mozilla Corporation users run:
```
./manage.py waffle_flag foxfood --group=mozilla_corporation --create
```
3. Send an email to a Mozilla user
4. Check that the email header now has another link with the text "Report Breakage" and looks like
Screen Shot 2022-04-25 at 15 46 57
6. Check that trackers in the general category was replaced with /faq link
7. 5. Check that logging was done and looks like:
```
{"Timestamp": 1650918098061819904, "Type": "studymetrics", "Logger": "fx-private-relay", "Hostname": "hostname", "EnvVersion": "2.0", "Severity": 6,"Fields": {"level": "general", "tracker_removed": 0, "general": {"count": 2, "trackers": {"open.tracker.com": 2}}, "strict": {"count": 0, "trackers": {}}, "msg": "email_tracker_foxfooding_summary"}}
```

## Add a user to the `foxfood` flag using an email results in user's email stripped from general tracker and adds breakage survey link
1. To add the user to existing `foxfood` flag, run:
```
./manage.py waffle_flag foxfood --append --user=email@address.com --group=mozilla_corporation
```
2. Send an email to the user's Relay address
3. Check that the email header now has another link with the text "Report Breakage" and looks like
Screen Shot 2022-04-25 at 15 46 57
4. Check that trackers in the general category was replaced with /faq link
5. Check that logging was done and looks like:
```
{"Timestamp": 1650918098061819904, "Type": "studymetrics", "Logger": "fx-private-relay", "Hostname": "hostname", "EnvVersion": "2.0", "Severity": 6,"Fields": {"level": "general", "tracker_removed": 0, "general": {"count": 2, "trackers": {"open.tracker.com": 2}}, "strict": {"count": 0, "trackers": {}}, "msg": "email_tracker_foxfooding_summary"}}
```

## Users not in foxfood flag gets email relayed as normal
1. Send email to relay for a user who is not a Mozillian and NOT listed for users in flag
2. Check that relayed email does not have a header to "Report Breakage"
3. Check that email is relayed as it has
4. `email_tracker_foxfooding_summary` was NOT logged

# Checklist

- [ ] l10n dependencies have been merged, if any.
- [ ] All acceptance criteria are met.
- [ ] I've added or updated relevant docs in the docs/ directory.
- [ ] All UI revisions follow the [coding standards](https://github.com/mozilla/fx-private-relay/blob/main/docs/coding-standards.md), and use Protocol tokens where applicable (see `/static/scss/libs/protocol/css/includes/tokens/dist/index.scss`).
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

# Follow-up
- Remove user from group
- User email address update triggers removal or addition to Foxfood flag
-  `update_user_group` command distinguishes between users that were already in the group and new users added to the group

### CODE REVIEW

**Review by jwhitlock — COMMENTED** [review]
Thanks @say-yawn! Here's some initial feedback.

File: emails/utils.py [hunk 1]
```diff
@@ -375,4 +390,29 @@ def count_all_trackers(html_content):
...
+        pattern = convert_domains_to_regex_patterns(tracker)
+        changed_content, matched = re.subn(pattern, f'\g<1>{settings.SITE_ORIGIN}/faq\g<1>', changed_content)
+        tracker_count += matched
+
+    
+    incr_if_enabled('tracker_foxfooding.general_count', tracker_count)
+    incr_if_enabled('tracker_foxfooding.strict_count', strict_detail['count'])
+    study_logger.info(
+        'email_tracker_foxfooding_summary',
+        extra={'general': general_detail, 'strict': strict_detail}
```
Comment (jwhitlock) [inline emails/utils.py:416]: Can we add `tracker_count` and `level` to the log as well?

File: emails/views.py [hunk 2]
```diff
@@ -457,12 +459,24 @@ def _sns_message(message_json):
...
 
     message_body = {}
+    email_tracker_study_link = ''
     if html_content:
         incr_if_enabled('email_with_html_content', 1)
+        foxfood_flag = Flag.objects.filter(name='foxfood').first()
+        if foxfood_flag and foxfood_flag.is_active_for_user(address.user):
+            html_content, removed_count, strict_count = remove_trackers(html_content)
+            email_tracker_study_link = (
+                'https://www.surveygizmo.com/s3/6837234/Relay-General-Email-Tracker-Removal-2022?'
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/views.py:469]: Updating this. Thanks for catching the default field.

[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/views.py:469]: Screen Shot 2022-04-25 at 12 43 35
Fixed

File: emails/utils.py [hunk 3]
```diff
@@ -375,4 +388,29 @@ def count_all_trackers(html_content):
...
         'email_tracker_summary',
         extra={'general': general_detail, 'strict': strict_detail}
-    )
\ No newline at end of file
+    )
+
+def convert_domains_to_regex_patterns(domain_pattern):
+    return '(["])(\\S*://\\S*.' + domain_pattern + '\\S*)\\1'
+
+def remove_trackers(html_content, level='general'):
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/utils.py:398]: resolved by f88af3e97fe74fdc2c30bd66890b4f5bfb8386f9

File: emails/utils.py (hunk 1 above)
Comment (say-yawn) [reply emails/utils.py]: done by 916082ab643b88179ed5eccba89d6aab82f411c7

File: emails/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/utils.py:398]: also resolved by 916082ab643b88179ed5eccba89d6aab82f411c7

File: emails/utils.py [hunk 4]
```diff
@@ -375,4 +388,29 @@ def count_all_trackers(html_content):
     study_logger.info(
         'email_tracker_summary',
         extra={'general': general_detail, 'strict': strict_detail}
-    )
\ No newline at end of file
+    )
+
+def convert_domains_to_regex_patterns(domain_pattern):
+    return '(["])(\\S*://\\S*.' + domain_pattern + '\\S*)\\1'
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/utils.py]: resolved by f88af3e97fe74fdc2c30bd66890b4f5bfb8386f9 with `re.escape(domain_pattern)`

File: emails/utils.py [hunk 5]
```diff
@@ -375,4 +388,29 @@ def count_all_trackers(html_content):
...
+        tracker_count += matched
+
+    
+    incr_if_enabled('tracker_foxfooding.general_count', tracker_count)
+    incr_if_enabled('tracker_foxfooding.strict_count', strict_detail['count'])
+    study_logger.info(
+        'email_tracker_foxfooding_summary',
+        extra={'general': general_detail, 'strict': strict_detail}
+    )
+    return changed_content, tracker_count, strict_detail['count']
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/utils.py]: resolved by: 916082ab643b88179ed5eccba89d6aab82f411c7

File: emails/utils.py [hunk 6]
```diff
@@ -353,7 +354,18 @@ def remove_message_from_s3(bucket, object_key):
...
 
-
+def set_user_group(user):
+    email_domain = user.email.split('@')[1]
+    group_attribute = {
+        'mozilla.com': 'mozilla_corporation',
+        'mozillafoundation.org': 'mozilla_foundation',
+        'getpocket.com': 'pocket'
+    }
+    internal_group_qs = Group.objects.filter(name=group_attribute.get(email_domain))
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/utils.py:367]: Done edc4d163c711924dd8beedf562aafd316aed5586

File: emails/management/commands/update_user_group.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        parser.add_argument('email_domain', nargs=1)
+
+    def handle(self, *args, **options):
+        email_domain = options['email_domain'][0]
+        user_qs = User.objects.filter(email__endswith=email_domain)
+        update_count = 0
+        for user in user_qs:
+            set_user_group(user)
+            update_count += 1
+        self.stdout.write(f"Updated {update_count} users group.")
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/management/commands/update_user_group.py]: Done edc4d163c711924dd8beedf562aafd316aed5586

File: emails/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/utils.py:367]: Created issue to track this recommendation: https://mozilla-hub.atlassian.net/browse/MPP-1919

File: emails/management/commands/update_user_group.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/management/commands/update_user_group.py]: Created issue to track this recommendation: https://mozilla-hub.atlassian.net/browse/MPP-1919

File: emails/tests/utils_tests.py [hunk 8]
```diff
@@ -103,3 +106,73 @@ def test_get_domains_from_settings(self):
...
+        assert strict_count == 0
+
+    def test_complex_single_quote_general_tracker_replaced_with_relay_content(self):
+        content = (
+            '<a href=\'https://foo.open.tracker.com/foo/bar.html\'>A link</a>\n'
+            + '<img src=\'https://bar.open.tracker.com/foo/bar.jpg\'>An image</img>'
+        )
+        changed_content, general_removed, general_count, strict_count = remove_trackers(content)
+
+        # assert changed_content == self.expected_content
```
Comment (jwhitlock) [inline emails/tests/utils_tests.py:151]:
```suggestion
        assert changed_content == self.expected_content.replace('"', "'")
```

File: emails/tests/utils_tests.py [hunk 9]
```diff
@@ -103,3 +106,73 @@ def test_get_domains_from_settings(self):
...
+    def test_no_tracker_replaced_with_relay_content(self):
+        content = (
+            '<a href="https://fooopen.tracker.com/foo/bar.html">A link</a>\n'
+            + '<img src="https://baropen.tracker.com/foo/bar.jpg">An image</img>'
+        )
+        changed_content, general_removed, general_count, strict_count = remove_trackers(content)
+
+        assert changed_content == content
+        assert general_removed == 0
+        assert general_count == 2  # this is because the count uses search and not regex pattern
```
Comment (jwhitlock) [inline emails/tests/utils_tests.py:165]: Thanks for this comment!

File: emails/tests/utils_tests.py [hunk 10]
```diff
@@ -103,3 +106,73 @@ def test_get_domains_from_settings(self):
...
+        assert strict_count == 0
+    
+    def test_complex_general_tracker_replaced_with_relay_content(self):
+        content = (
+            '<a href="https://foo.open.tracker.com/foo/bar.html">A link</a>\n'
+            + '<img src="https://bar.open.tracker.com/foo/bar.jpg">An image</img>'
+        )
+        changed_content, general_removed, general_count, strict_count = remove_trackers(content)
+
+        # assert changed_content == self.expected_content
```
Comment (jwhitlock) [inline emails/tests/utils_tests.py:139]: This assertion works, can be uncommented.
```suggestion
        assert changed_content == self.expected_content
```

File: emails/tests/utils_tests.py [hunk 11]
```diff
@@ -103,3 +106,73 @@ def test_get_domains_from_settings(self):
...
         }
+
+@override_settings(SITE_ORIGIN='https://test.com')
+@patch('emails.utils.GENERAL_TRACKERS', ['open.tracker.com'])
+@patch('emails.utils.STRICT_TRACKERS', ['strict.tracker.com'])
+class RemoveTrackers(TestCase):
+    def setUp(self):
+        self.expected_content = (
+            '<a href="https://test.com/faq">A link</a>\n'
+            + '<img src="https://test.com/faq">An image</img>'
```
Comment (jwhitlock) [inline emails/tests/utils_tests.py:117]: We may need to be more clever for the next version about replacing links to images, but this does get rid of the tracker.

**Review by jwhitlock — CHANGES_REQUESTED** [review]
Thanks @say-yawn, I appreciate the changes! A few changes:

* Add content assertions for `test_complex_general_tracker_replaced_with_relay_content` and `test_complex_single_quote_general_tracker_replaced_with_relay_content`
* Avoid a database query for non-Mozilla emails

File: emails/tests/utils_tests.py (hunk 11 above)
Comment (say-yawn) [reply emails/tests/utils_tests.py:117]: Added as a follow-up item https://mozilla-hub.atlassian.net/browse/MPP-1919

**Review by jwhitlock — APPROVED** [review]
Thanks @say-yawn! The testing instructions worked with my local environment. I get two logs when forwarding an email:

```
{"Timestamp": 1650922382662334976, "Type": "studymetrics", "Logger": "fx-private-relay", "Hostname": "johnwhitlock-4zq05n", "EnvVersion": "2.0", "Severity": 6, "Pid": 43025, "Fields": {"level": "general", "tracker_removed": 20, "general": {"count": 21, "trackers": {"acemlna.com": 19, "activehosted.com": 2}}, "strict": {"count": 0, "trackers": {}}, "msg": "email_tracker_foxfooding_summary"}}
{"Timestamp": 1650922383406768896, "Type": "studymetrics", "Logger": "fx-private-relay", "Hostname": "johnwhitlock-4zq05n", "EnvVersion": "2.0", "Severity": 6, "Pid": 43025, "Fields": {"general": {"count": 0, "trackers": {}}, "strict": {"count": 0, "trackers": {}}, "msg": "email_tracker_summary"}}
```

I think that means ``email_tracker_summary`` will represent emails _after_ trackers are removed.

The email was forwarded with all the content links changed to links to our FAQ. This broken the email, which was a list of links to articles. There was a "Report Breakage" link in the header, and I was able to follow it and fill out the survey.

**Comment by jwhitlock** [comment]
Thanks, that makes the logs more useful:

```
{"Timestamp": 1650924660676921088, "Type": "studymetrics", "Logger": "fx-private-relay", "Hostname": "johnwhitlock-4zq05n", "EnvVersion": "2.0", "Severity": 6, "Pid": 43974, "Fields": {"general": {"count": 22, "trackers": {"acemlna.com": 20, "activehosted.com": 2}}, "strict": {"count": 0, "trackers": {}}, "msg": "email_tracker_summary"}}
{"Timestamp": 1650924660897246976, "Type": "studymetrics", "Logger": "fx-private-relay", "Hostname": "johnwhitlock-4zq05n", "EnvVersion": "2.0", "Severity": 6, "Pid": 43974, "Fields": {"level": "general", "tracker_removed": 21, "general": {"count": 22, "trackers": {"acemlna.com": 20, "activehosted.com": 2}}, "strict": {"count": 0, "trackers": {}}, "msg": "email_tracker_foxfooding_summary"}}
{"Timestamp": 1650924661842960896, "Type": "request.summary", "Logger": "fx-private-relay", "Hostname": "johnwhitlock-4zq05n", "EnvVersion": "2.0", "Severity": 6, "Pid": 43974, "Fields": {"errno": 0, "agent": "Amazon Simple Notification Service Agent", "lang": "", "method": "POST", "path": "/emails/sns-inbound", "uid": "", "rid": "4dde6ed8-f452-4fda-b58b-ea8052e192dc", "t": 1921}}
```

**Pull request merged**